File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        '''
        0, 1, 2
        1, 2, 3
           
        f(i = 0, vals = {1, 2, 3}, arr = [])
            f(1, {2, 3}, [1])
                f(2, {3}, [1, 2])
                    f(3, {}, [1, 2, 3]) *
                f(2, {2}, [1, 3])
            f(1, {1, 3}, arr = [2])
        
        '''
        # res = [1, 3, 5]
        # res.insert(2, 2)
        # print(res)
        
        res = []
        n = len(nums)
        
        def backtracking(i, vals, arr):
            if i == n:
                res.append(copy.deepcopy(arr))
                return
            for j in range(len(vals)):
                num = vals.pop(j)
                arr.append(num)
                backtracking(i + 1, vals, arr)
                vals.insert(j, num)
                arr.pop()
                
        backtracking(0, nums, [])
                
        
        return res

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        combo = nums;
        backtrack(0);
        return result;
    }

private:
    vector<vector<int>> result;
    vector<int> combo;

    void swap(int i, int j) {
        int temp = combo[i];
        combo[i] = combo[j];
        combo[j] = temp;
    }
    
    void backtrack(int start) {
        if (start >= combo.size()) {
            vector<int> copy = combo;
            result.push_back(copy);
            return;
        }

        for (int i = start; i < combo.size(); i++) {
            swap(start, i);
            backtrack(start + 1);
            swap(start, i);
        }
    }
};

/*
0, 1, 2
1, 2, 3

f(start = 0)
    i = 0: 
        swap (0, 0) -> [1, 2, 3]
        f(start = 1)
            i = 1: 
                swap (1, 1) -> [1, 2, 3]
                f(start = 2)
                    i = 2: 
                        swap (2, 2) -> [1, 2, 3]
                        f(start = 3) -> **[1, 2, 3]
                        back (2, 2) -> [1, 2, 3]
                back (1, 1)
            i = 2:
                swap (1, 2) -> [1, 3, 2]
                f(start = 3) -> **[1, 3, 3]
                back (2, 1) -> [1, 2, 3]
        back (0, 0)
    i = 1:
        swap (0, 1) -> [2, 1, 3]
        f(start + 1 = 1)
            i = 1:
                swap (1, 1) -> [2, 1, 3]
                f(start = 2)
                    swap (2, 2) -> [2, 1, 3]
                    f(start = 3) -> **[2, 1, 3]
                    back (2, 2)
                back (1, 1)
            i = 2:
                swap (1, 2) -> [2, 3, 1]

*/